[compiler][api] Enable custom registration of Relay ops from python

bwasti opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#42 [compiler][api] Enable custom registration of Relay ops from python**

bwasti committed:
[compiler][api] Enable custom registration of Relay ops from python

bwasti committed:
Update on "[compiler][api] Enable custom registration of Relay ops from python"

[compiler][api] Enable custom registration of Relay ops from python

gh-metadata: pytorch tvm 42 gh/bwasti/26/head

wanchaol commented on torch_tvm/register.cpp:
```diff
@@ -22,55 +23,93 @@ static std::string device_type = "cpu";
+    AT_ASSERT(relay_op);
+    std::vector<Argument> torch_inputs;
+    Graph wrapper_graph;
+    std::vector<Value *> graph_inputs;
```
nit: `Value*`?

wanchaol commented on torch_tvm/register.cpp:
```diff
@@ -22,55 +23,93 @@ static std::string device_type = "cpu";
+          }}});
+
+    auto relay_op = tvm::relay::Op::Get(relay_name);
+    AT_ASSERT(relay_op);
```
this should be `AT_CHECK` and send msg to user if there's no such op exist?

wanchaol commented on torch_tvm/register.cpp:
```diff
@@ -22,55 +23,93 @@ static std::string device_type = "cpu";
+          [relay_name](Node *node, tvm::Array<tvm::relay::Expr> inputs) {
+            auto op = tvm::relay::Op::Get(relay_name);
+            AT_ASSERT(op);
+            return tvm::relay::CallNode::make(op, inputs, tvm::Attrs(), {});
```
does this mean that we currently can only register simple operators that does not require custom Attrs?

bwasti replied:
yep, unfortunately.  There isn't a great way to expose the relay Attrs interface to python (as far as I can tell)

wanchaol commented on torch_tvm/register.cpp:
```diff
@@ -22,55 +23,93 @@ static std::string device_type = "cpu";
+    wrapper_graph.registerOutput(node->output());
+    // Convert node to a tvm compilation group containing the node
+    node = SubgraphUtils::createSingletonSubgraph(node, tvm_sym);
+    auto cc = std::make_shared<TVMCompiler>(node, opt_level, strict,
```
from what I observed here, for each custom operators, we are creating a single separate `TVMCompiler` group and execute them separately? will the `customFusion` pass take the compilation group and merge it with the bigger one?

bwasti replied:
customFusion will never be run on these operators because this interface is for eager mode only

wanchaol commented on test/test_integration.py:
```diff
@@ -0,0 +1,32 @@
+        
+        X = torch.randn(128,128)
+        print(test.graph)
+        print(test.graph_for(X,X))
```
it is hard to see the graph from the PR, why don't we just call `runBoth` here?

bwasti committed:
Update on "[compiler][api] Enable custom registration of Relay ops from python"

[compiler][api] Enable custom registration of Relay ops from python

gh-metadata: pytorch tvm 42 gh/bwasti/26/head

bwasti committed:
Update on "[compiler][api] Enable custom registration of Relay ops from python"

[compiler][api] Enable custom registration of Relay ops from python

gh-metadata: pytorch tvm 42 gh/bwasti/26/head

wanchaol reviewed (approved):
looks good.

bwasti commented:
There are some issues with this API that make it pretty useless, so I am going to abandon it.  That being said, there are some tricks here worth using
